Enable DataParallel to run zero input Module

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46565 Enable DataParallel to run zero input Module**

This is to fix #44873, allow zero input module in DataParallel. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

Commit:
error out if DataParallel has zero input but specified device ids

[ghstack-poisoned]

Commit:
Update on "error out if DataParallel has zero input but specified device ids"


This is to fix #44873, prevent crash. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

Commit:
Update on "error out if DataParallel has zero input but specified device ids"


This is to fix #44873, preventing crash. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/data_parallel.py:
```diff
@@ -148,6 +148,16 @@ def forward(self, *inputs, **kwargs):
+        # dict has item. Caller needs to make sure 'kwargs' dict has at least an input for
+        # forward function in this case.
+        if len(inputs) == 0 and len(kwargs) == 0:
+            raise RuntimeError("Forward function must have at least one input, bot got zero")
```
This looks OK to me.

But the use case mentioned in that issue looks valid to me, so I wonder if we should fix that use case instead of error out? 

Let me know what do you think, and whether you would like to fix. I can stamp this one if you think fixing is out of the scope. 

To fix, we need to following:

1. The segment fault occurs on `self.replicate`, we need to add the following to `replicate.py`:
    ```python
    if not devices:
        return []
    ```
2. After fixing that, we also need to update `parallel_apply.py` and `scatter_gather.py` accordingly to handle empty modules and empty outputs respectively.

glaringlee replied:
per our discussion, I will fix this entirely.

Commit:
Update on "error out if DataParallel has zero input but specified device ids"


This is to fix #44873, preventing crash. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

Commit:
Update on "error out if DataParallel has zero input but specified device ids"


This is to fix #44873, preventing crash. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

glaringlee commented:
@mrshenli 
I revisited the entire logic from scatter->replicate->parallel_apply->gather.
I think this is the simplest and cleanest fix at this point.

When inputs is empty, there is actually nothing to parallel based on current design. There are two paths after change:

1. Call module() directly as what we do for single device here:
https://github.com/pytorch/pytorch/blob/0910c26c6453a86ef3623fba1718029701a17539/torch/nn/parallel/data_parallel.py#L165
2. Still do scatter, gather process when multiple device ids are given, but only run module on 1 device:
I am not sure whether this is meaningful, but I found that current DataParallel do support such case: say your module is on cuda:1, but your forward function generates output (tensor) that on cuda:0 and you want the output on cuda:1 as well. What you can do is that passing device_ids as [1, 0]. So this will bypass calling module() directly, and have the output on cuda:1. I think the DataParallel class is not that consistent at this point, so I prefer to keep this behavior instead of always calling module() directly.

Both paths need the inputs and kwargs has an empty list/dict. So I added it and everything works fine. replicate() shouldn't crash when devices is empty, so I added a sanity check there as well. Let me know if this works for you.

Commit:
Update on "Enable DataParallel to run zero input Module"


This is to fix #44873, allow zero input module in DataParallel. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

Commit:
Update on "Enable DataParallel to run zero input Module"


This is to fix #44873, allow zero input module in DataParallel. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

Commit:
Update on "Enable DataParallel to run zero input Module"


This is to fix #44873, allow zero input module in DataParallel. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

Commit:
Update on "Enable DataParallel to run zero input Module"


This is to fix #44873, allow zero input module in DataParallel. But we do recommend DDP instead of DataParallel.

Differential Revision: [D24405275](https://our.internmc.facebook.com/intern/diff/D24405275)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/replicate.py:
```diff
@@ -80,6 +80,9 @@ def replicate(network, devices, detach=False):
         raise RuntimeError("Cannot replicate network where python modules are "
                            "childrens of ScriptModule")
 
+    if not devices:
```
do we still need this? Looks like there is no valid way to trigger this as `self.device_ids[:len(inputs)]` will at least contain one device?

glaringlee replied:
Prefer to keep this. It is a sanity check to prevent potential crash when devices is empty, though it won't be trigger for now.

mrshenli reviewed (approved):
Great fix! Thanks!
